### ISSUE TITLE
contrastratio

### ISSUE THREAD

**Issue by Kerruba** [brehaut/color-js#21]
Hello, 

I was wondering what you think about integrating in the model also a "colorContrast" method that check the color contrast between two colors.

**Comment by brehaut** [comment]
That seems like a sensible function to have. I don't have time to implement it myself at present, but a pull request would be welcome.

**Comment by brehaut** [comment]
Some notes on implementation based on information from https://www.w3.org/TR/WCAG/#visual-audio-contrast the relevant links for defining contrast ratios wrt to web standards:

- https://www.w3.org/TR/WCAG/#contrast-ratiodef 
- https://www.w3.org/TR/WCAG/#relativeluminancedef

Any pull requests for this feature would need to satisfy these definitions. 

These recommendations look like they should be implemented on the RGB model, unless it can be shown that the HSL luminance calculation is equivalent to the sRGB one from w3.

**Comment by Kerruba** [comment]
I suppose we can replicate what Lea Verou as already made in her contrast-ratio project.
https://github.com/LeaVerou/contrast-ratio/blob/gh-pages/color.js#L108

**Comment by Kerruba** [comment]
Yeah, I couldn't find anything related to HSL luminance...probably the only way to integrate this into also that model is to do an inner conversion to RGB and calculate on that the contrast-ratio. But doesn't look really elegant, even if it will work

**Comment by brehaut** [comment]
It looks like a reasonable starting point; If nothing else it would prove a useful for testing correctness.

**Comment by brehaut** [comment]
Conversion to RGB is trivial; just dont implement the contrastRatio method on anything but the RGB model. stupid metaprogramming magic will take care of the conversion implicitly†. The RGB model is the correct choice here simply because it means that the implementation of the algorithm can trivially be compared to the definition of the algorithm for ensuring correctness. 

Regarding the HSL Luminance, it's just the L value so it would be a simple case of implementing the ratio calculation against that value. I think there's a separate issue here where the various conversions and calculations for the different models and values should be checked against the W3C definitions for those models in CSS. Ideally computing contrast ratios could be implemented on either of RGB or HSL and produce consistent results. This is one area where the project really shows its age (it largely predates CSS supporting a lot of this stuff).

† The magic is a little brittle when adding new models but maintaining the existing ones is fine. But implicit conversions is half the reason I wrote this originally; some operations just suit some models better. I would kind of like to make the models be used to define a graph and then calculate shortest paths to determine the conversions but that's fairly overkill.

**Comment by Kerruba** [comment]
> Regarding the HSL Luminance, it's just the L value so it would be a simple case of implementing the ratio calculation against that value.
Not so sure about this, I've tried some values on the LeaVerou (assuming is correct) and here the results:
- hsl(360,80%,30%) -> luminance: 0.10779410842543202
- hsl(360,80%,90%) -> luminance: 0.7552837134836996
- hsl(150,30%,90%) -> luminance: 0.8739183779656772

**Comment by brehaut** [comment]
It's entirely plausible that the CSS specifications for HSL and RGB suffer from the same mapping concerns as color-js ;)
